WrapImage/nifti_wrapper.py

```diff
@@ -0,0 +1,127 @@
+
+    if not os.path.exists(json_file):
+        raise FileNotFoundError(f"File '{json_file}' not found.")
+
+    with open(json_file, "r") as f:
+        try:
+            json_data = json.load(f)
+        except json.JSONDecodeError as e:
+            raise ValueError(f"Error decoding JSON in file '{json_file}': {e}")
+
+    return json_data
+
+def read_bval_file(bval_file):
+    """
+    For reading the bval file
+    """
+    if not os.path.exists(bval_file):
+        raise FileNotFoundError(f"File '{bval_file}' not found.")
+
+    bval_data = np.genfromtxt(bval_file, dtype=float)
+    return bval_data
+
+def read_bvec_file(bvec_file):
+    """
+    For reading the bvec file
+    """
+    if not os.path.exists(bvec_file):
+        raise FileNotFoundError(f"File '{bvec_file}' not found.")
+
+    bvec_data = np.genfromtxt(bvec_file)
+    bvec_data = np.transpose(bvec_data)  # Transpose the array
+    return bvec_data
+
+def save_nifti_file(data, output_file, affine=None, **kwargs):
+    """
+    For saving the 3d nifti images of the output of the algorithm
+    """
+    if affine is None:
+        affine = np.eye(data.ndim + 1)
+    output_img = nib.nifti1.Nifti1Image(data, affine , **kwargs)
+    nib.save(output_img, output_file)
+
+def loop_over_first_n_minus_1_dimensions(arr):
+    """
+    Loops over the first n-1 dimensions of a numpy array.
+
+    Args:
+        arr: A numpy array.
+
+    Yields:
+        A tuple containing the indices for the current iteration and a flattened view of the remaining dimensions.
+    """
+    n = arr.ndim
+    for idx in np.ndindex(*arr.shape[:n-1]):
+        flat_view = arr[idx].flatten()
+        yield idx, flat_view
+
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Read a 4D NIfTI phantom file along with BIDS JSON, b-vector, and b-value files.")
+    parser.add_argument("input_file", type=str, help="Path to the input 4D NIfTI file.")
+    parser.add_argument("bvec_file", type=str, help="Path to the b-vector file.")
+    parser.add_argument("bval_file", type=str, help="Path to the b-value file.")
+    parser.add_argument("--affine", type=float, nargs="+", help="Affine matrix for NIfTI image.")
+    parser.add_argument("--algorithm", type=str, choices=["algorithm1", "algorithm2"], default="OJ_GU_seg", help="Select the algorithm to use.")
+    parser.add_argument("algorithm_args", nargs=argparse.REMAINDER, help="Additional arguments for the algorithm.")
+
+    args = parser.parse_args()
+
+    try:
+        # Read the 4D NIfTI file
+        data, _ = read_nifti_file(args.input_file)
+
+        # Read the b-vector, and b-value files
+        bvecs = read_bvec_file(args.bvec_file)
+        bvals = read_bval_file(args.bval_file)
+
+        # Pass additional arguments to the algorithm
+
+        fit = OsipiBase(algorithm=args.algorithm)
+        f_image = []
+        Dp_image = []
+        D_image = []
+
+        for idx, view in loop_over_first_n_minus_1_dimensions(data):
```

Comment:
I tried it and didn't realize it was fitting and cancelled the run only to realize it was actually working correctly. So perhaps wrap this loop in `tqdm`? https://github.com/tqdm/tqdm

It's already a requirement so should be a small addition. If that's harder than expected, just a print before the loop saying that it's fitting with the given algorithm would be good enough too.